perception/image_projection_based_fusion/src/pointpainting_fusion/node.cpp

```diff
@@ -163,77 +252,77 @@ void PointPaintingFusionNode::fuseOnSingleImage(
-    camera_info.p.at(3), camera_info.p.at(4), camera_info.p.at(5), camera_info.p.at(6),
-    camera_info.p.at(7), camera_info.p.at(8), camera_info.p.at(9), camera_info.p.at(10),
-    camera_info.p.at(11);
-
   // transform
   sensor_msgs::msg::PointCloud2 transformed_pointcloud;
   tf2::doTransform(painted_pointcloud_msg, transformed_pointcloud, transform_stamped);
 
+  std::chrono::system_clock::time_point start, end;
+  start = std::chrono::system_clock::now();
+
+  const auto x_offset =
+    transformed_pointcloud.fields.at(static_cast<size_t>(autoware_point_types::PointIndex::X))
+      .offset;
+  const auto y_offset =
+    transformed_pointcloud.fields.at(static_cast<size_t>(autoware_point_types::PointIndex::Y))
+      .offset;
+  const auto z_offset =
+    transformed_pointcloud.fields.at(static_cast<size_t>(autoware_point_types::PointIndex::Z))
+      .offset;
+  const auto class_offset = painted_pointcloud_msg.fields.at(4).offset;
+  const auto p_step = transformed_pointcloud.point_step;
+  // projection matrix
+  Eigen::Matrix3f camera_projection;  // use only x,y,z
+  camera_projection << camera_info.p.at(0), camera_info.p.at(1), camera_info.p.at(2),
+    camera_info.p.at(4), camera_info.p.at(5), camera_info.p.at(6);
+  /** dc : don't care
+
+x    | f  x1 x2  dc ||xc|
+y  = | y1  f y2  dc ||yc|
+dc   | dc dc dc  dc ||zc|
+                     |dc|
+   **/
+
+  auto objects = input_roi_msg.feature_objects;
+  int iterations = transformed_pointcloud.data.size() / transformed_pointcloud.point_step;
   // iterate points
-  sensor_msgs::PointCloud2Iterator<float> iter_car(painted_pointcloud_msg, "CAR");
-  sensor_msgs::PointCloud2Iterator<float> iter_ped(painted_pointcloud_msg, "PEDESTRIAN");
-  sensor_msgs::PointCloud2Iterator<float> iter_bic(painted_pointcloud_msg, "BICYCLE");
-
-  for (sensor_msgs::PointCloud2ConstIterator<float> iter_x(transformed_pointcloud, "x"),
-       iter_y(transformed_pointcloud, "y"), iter_z(transformed_pointcloud, "z");
-       iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z, ++iter_car, ++iter_ped, ++iter_bic) {
-    // filter the points outside of the horizontal field of view
-    if (
-      *iter_z <= 0.0 || (*iter_x / *iter_z) > tan_h_.at(image_id) ||
-      (*iter_x / *iter_z) < -tan_h_.at(image_id)) {
+  // Requires 'OMP_NUM_THREADS=N'
+  omp_set_num_threads(omp_num_threads_);
+#pragma omp parallel for
+  for (int i = 0; i < iterations; i++) {
+    int stride = p_step * i;
+    unsigned char * data = &transformed_pointcloud.data[0];
+    unsigned char * output = &painted_pointcloud_msg.data[0];
+    float p_x = *reinterpret_cast<const float *>(&data[stride + x_offset]);
+    float p_y = *reinterpret_cast<const float *>(&data[stride + y_offset]);
+    float p_z = *reinterpret_cast<const float *>(&data[stride + z_offset]);
+    if (p_z <= 0.0 || (p_x / p_z) > tan_h_.at(image_id) || (p_x / p_z) < -tan_h_.at(image_id)) {
```

Comment:
How about changing `(p_x / p_z) > tan_h_.at(image_id)` to `p_x > (tan_h_.at(image_id) * p_z)` to avoid division?

Reply:
thanks for the suggestion. https://github.com/autowarefoundation/autoware.universe/pull/4314/commits/a5c023f2b35dce04121d9ea75535189999e8bf0d#diff-e8df78707e8cee5e869c96c349aca84fa2126bc70e25f093bf10f8a325e638e2R298